### PR TITLE
Custom fields

### PR DESCRIPTION
**Pull request by mkurek**
Custom fields described in #2158 

TODO:
- [x] functional testing (both gui and api)
- [x] simplify attaching custom fields to api resource
- [x] fix unique validation (https://code.djangoproject.com/ticket/12028)
- [x] docs
- [x] merge migrations

### CODE REVIEW

**Comment by mkurek** [comment]
ralph_custom_fields2

**Comment by vi4m** [comment]
Fucking awesome code :-)

**Comment by ar4s** [comment]
:ok:

File: src/ralph/urls/test.py [hunk 1]
```diff
@@ -2,14 +2,17 @@
...
     urls as perm_view_url  # noqa
 from ralph.urls.base import urlpatterns as base_urlpatterns
 
+
 urlpatterns = base_urlpatterns
 urlpatterns += [
     url(r'^', include(lib_api.urlpatterns)),
     url(r'^', include(perm_view_url)),
     url(r'^', include(ralph_api.urlpatterns)),
+    url(r'^', include(custom_fields_tests_urls.urlpatterns)),
```
Comment (ghost) [inline src/ralph/urls/test.py:17]: I can't see defined urlpatterns in `ralph.lib.custom_fields.tests`

File: src/ralph/lib/custom_fields/models.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            })
+        return field_type.form_field(**params)
+
+
+class CustomFieldValue(TimeStampMixin, models.Model):
+    custom_field = models.ForeignKey(
+        CustomField, verbose_name=_('key'), on_delete=models.PROTECT
+    )
+    # value is stored in charfield on purpose - ralph's custom field mechanism
+    # is by-design simple, so it, for example, don't allow to filter by range
```
Comment (ghost) [inline src/ralph/lib/custom_fields/models.py]: don't -> doesn't?

File: src/ralph/lib/custom_fields/models.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    content_type = models.ForeignKey(ContentType)
+    object_id = models.PositiveIntegerField()
+    object = generic.GenericForeignKey('content_type', 'object_id')
+
+    class Meta:
+        unique_together = ('custom_field', 'content_type', 'object_id')
+
+    def __str__(self):
+        return '{} ({}): {}'.format(
+            self.custom_field if self.custom_field_id else None,
```
Comment (ghost) [inline src/ralph/lib/custom_fields/models.py:113]: Is it possible that custom_field is missing?

Comment (mkurek) [inline src/ralph/lib/custom_fields/models.py:113]: no - custom_field is obligatory

File: src/ralph/lib/custom_fields/tests/admin.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+from django.contrib import admin
+
+from ..admin import CustomFieldValueAdminMaxin
```
Comment (ghost) [inline src/ralph/lib/custom_fields/tests/admin.py]: CustomFieldValueAdminMAXIN? ;)

**Comment by szok** [comment]
:ok:

File: src/ralph/urls/test.py (hunk 1 above)
Comment (mkurek) [inline src/ralph/urls/test.py:17]: https://github.com/allegro/ralph/pull/2426/files#diff-0a710b9a1ab06193c50cd4e7bee9faa3R1

**Comment by ghost** [comment]
:ok:

**Pull request merged**